### ISSUE TITLE
NSE Loss Function

### ISSUE THREAD

**Issue by rro2q2** [neuralhydrology/neuralhydrology#96]
Hello,

When computing the loss function `_get_loss()` in the MaskedNSELoss class, could you explain why you performed standard deviation over the observations from a particular basin and then applied it to the numerator? From my understanding of calculating std, aren't you computing the square root of the squared error divided by the number of observations? 

i.e., sqrt(sum(sq_err) /  N), where N is the number of observations.

But this seems like a mismatch from the NSE loss equation in the [Nash and Sutcliffe (1970)](https://www.sciencedirect.com/science/article/abs/pii/0022169470902556) paper, where the denominator is the squared error of the observed streamflow minus the mean of the observed streamflow. 

i.e., 1 - ( (sum(Q_m[t] - Q_o[t]))**2 / (sum(Q_o[t] - Q_bar_o[t])**2)

What happens to the N from the std calculation? Is it used as a scaled factor for computing the loss function? I hope this all makes sense.

**Comment by kratzert** [comment]
Hi @rro2q2 

The loss that we are computing is described in https://hess.copernicus.org/articles/23/5089/2019/#section2 in Section 2.3 "Objective function: a smooth-joint NSE", as also indicated in the documentation of the MaskedNSELoss. Did you have a look at it?

This loss is an adapted version to:
a) account for the fact of stochastic mini-batch training and
b) for the fact of random samples from random basins in the multi-basin setting.

If your batch would be the entire training period of one basin, then this loss would be identical to the NSE as defined by Nash and Sutcliffe (with the difference of a fixed scalar because of the mean operator and the epsilon).

**Comment by rro2q2** [comment]
Hi @kratzert. Thanks for providing this resource, as this makes things clearer 👍
